oak-jcr/src/main/java/org/apache/jackrabbit/oak/jcr/delegate/SessionDelegate.java

```diff
@@ -78,6 +79,23 @@ public class SessionDelegate {
     static final Logger readOperationLogger = LoggerFactory.getLogger("org.apache.jackrabbit.oak.jcr.operations.reads");
     static final Logger writeOperationLogger = LoggerFactory.getLogger("org.apache.jackrabbit.oak.jcr.operations.writes");
 
+    // the bitmask used for trace level logging
+    // we use a bitmask instead of a counter to avoid the slow modulo operation:
+    // https://stackoverflow.com/questions/27977834/why-is-modulus-operator-slow
+    // so we use "if ((counter & LOG_TRACE_BIT_MASK) == 0) log(...)"
+    // instead of the slower "if ((counter % LOG_TRACE) == 0) log(...)"
+    // that means the values need to be some power of two, minus one
+    // log every 128th call by default
+    private static final long LOG_TRACE_BIT_MASK = Long.getLong(
+            "org.apache.jackrabbit.oak.jcr.operations.bitMask",
+            128L - 1);
+    // log a stack trace every ~1 million calls by default
+    private static final long LOG_TRACE_STACK_BIT_MASK = Long.getLong(
+            "org.apache.jackrabbit.oak.jcr.operations.stack.bitMask",
+            1024L * 1024 - 1);
+    // the counter used for logging
+    private static final AtomicLong LOG_COUNTER = new AtomicLong();
```

Comment:
What will happen if the counter overflows? I guess the bit shift operations will continue giving the expected results, but just want to be sure you considered this case. Not sure how long it will take to overflow a long, might be one of those age-of-universe situations, but better be sure.

Reply:
Yes, it will continue to work. Modulo, on the other hand, behaves in a slightly weird way for negative numbers: you get negative reminders. But checking for (x % y) == 0 would still work fine.